Speech and Braille: don't report many clickables in a row in browse mode.

michaelDCurran opened the pull request:
<!--
Please fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/wiki/Github-pull-request-template-explanation-and-examples
-->

### Link to issue number:
Fixes #7430

### Summary of the issue:
NVDA reports certain fields in browse mode as 'clickable' if the author has provided an onclick handler and NVDA would not normally report anything else for that field (E.g. has no reportable role).
However, with the increased use of onclick handlers on web pages, there are many cases where NVDA will report multiple clickables in a row E.g. "clickable clickable clickable clickable". This is extremely annoying.
To make things worse, versions of Chrome earlier than 63 also propagated the existence of the click action to all descendant accessibles, meaning that NVDA would report clickable for every descendant of a clickable node. Although this was fixed in Chrome a while back, many 3rd party apps that use Chromium / Electron (such as Microsoft Teams) still have this issue as they are using an older version of Chromium.
 
### Description of how this pull request fixes the issue:
Several solutions were described in issue #7430. This PR goes with solution 3 (collapsing clickables).
Although it wasn't the most perfect solution, it was the easiest to implement, and probably the easiest to follow code-wise. It also hopefully provides a good balance between usability and predictability. It is certainly a major improvement over what we currently have, but also erring on the side of reporting more than less.
In short: in both speech and braille now, NVDA will never report multiple clickables in a row without any other intervening text. As NVDA enters one or more clickable fields, it will only report clickable once. However, if one or more of these fields are exited, and then 1 or more new clickable fields are entered, NVDA will again report clickable, but only once when entering those new fields. To put it another way, NVDA will only report clickable once for any run of field starts.

 ### Testing performed:
Tested example in #7430. Multiple clickables in a row are now only announced as one clickable.
Also tested 
[clickables.html.txt](https://github.com/nvaccess/nvda/files/2551843/clickables.html.txt)
which contains various different paragraphs and divs containing single and nested clickable spans.

### Known issues with pull request:
None known.

### Change log entry:
Bug fixes:
* NVDA no longer reports clickable multiple times in a row when navigating clickable content in Browse Mode. (#7430)

michaelDCurran committed:
Speech and Braille: don't report many clickables in a row.

leonardder commented on source/braille.py:
```diff
@@ -872,7 +870,21 @@ def _addTextWithFields(self, info, formatConfig, isSelection=False):
+							if not inClickable:
+								# We have entered an outer most clickable or entered a new clickable after exiting a previous one 
+								# Report it if there is nothing else interesting about the field, but not if the user turned it off. 
+								if formatConfig['reportClickable']:
```
Would it be possible to check this earlier on? I guess it is not necessary to check the clickable state if we don't report clickables at all.

leonardder commented on source/speech.py:
```diff
@@ -853,9 +853,20 @@ def speakTextInfo(info,useCache=True,formatConfig=None,unit=None,reason=controlT
 				isTextBlank=False
 				_speakTextInfo_addMath(speechSequence,info,field)
 
+	# When true, we are inside a clickable field, and should therefore not announce any more new clickable fields
```
Could you also provide this comment in braille.py when declaring inClickable?

leonardder commented on source/speech.py:
```diff
@@ -853,9 +853,20 @@ def speakTextInfo(info,useCache=True,formatConfig=None,unit=None,reason=controlT
+			states=field.get('states')
+			if states and controlTypes.STATE_CLICKABLE in states:
+				# We entered the most outer clickable, so announce it, if we won't be announcing anything else interesting for this field, but not if the user turned them off.
+				if formatConfig['reportClickable']:
```
Same question as for de braille module.

leonardder commented on source/speech.py:
```diff
@@ -915,13 +928,30 @@ def speakTextInfo(info,useCache=True,formatConfig=None,unit=None,reason=controlT
+					if not inClickable:
+						# We have entered an outer most clickable or entered a new clickable after exiting a previous one 
+						# Announce it if there is nothing else interesting about the field, but not if the user turned it off. 
+						if formatConfig['reportClickable']:
```
Same concern as above.

michaelDCurran committed:
Address review actions and also simplify logic and fix a corner case.
Now we simply report clickable for the first clickable in a run of control starts. Hitting text, or control ends resets the run.

michaelDCurran commented:
@leonardder: I have addressed your review comments, and also somewhat 
simplified the logic, which actually fixes a particular corner case 
which was breaking.

Now we simply report the first clickable in a run of control starts, and 
text or a control end breaks the run.

Now all tests in clickables.html in both speech and braille work as 
expected. Previously the "Clickable spans in middle of clickable 
paragraph" test was not correct in braille.

derekriemer commented on source/speech.py:
```diff
@@ -915,9 +930,24 @@ def speakTextInfo(info,useCache=True,formatConfig=None,unit=None,reason=controlT
+						# Announce it if there is nothing else interesting about the field, but not if the user turned it off. 
+						presCat=command.field.getPresentationCategory(newControlFieldStack[0:],formatConfig,reason)
+						if not presCat or presCat is command.field.PRESCAT_LAYOUT:
+							tempTextList.append(controlTypes.stateLabels[controlTypes.STATE_CLICKABLE])
```
What if we keep a list of roles where all of that type of role is inherently clickable. That way we don't say something is a clibkable link or a clickable button? Or does this code already do that?

michaelDCurran replied:
We are already doing this by checking the field's presentationCategory.

derekriemer commented on source/speech.py:
```diff
@@ -915,9 +930,24 @@ def speakTextInfo(info,useCache=True,formatConfig=None,unit=None,reason=controlT
+						inClickable=True
+				text=info.getControlFieldSpeech(command.field,newControlFieldStack,"start_relative",formatConfig,extraDetail,reason=reason)
+				if text:
+					tempTextList.append(text)
```
Should text being true turn a run of clickables off, or do I misunderstand this? I'm reading it pretty fast.

michaelDCurran replied:
Nope. That text variable just holds  the speech for that particular control Start.

LeonarddeR reviewed (approved):
Looks way cleaner now, thanks!

michaelDCurran committed:
Update what's new.
